Root README.md improved for pkg.go.dev (#210)
canonical/ubuntu-image: README.md

The `gadget.yaml` file contains important information about the volumes that will be created. You can think of these volumes as disk images. The file also specifies the structures that will be included within each volume. Structures define the layout of the volume, including partitions, Master Boot Records, and any other relevant content. Also, the `gadget.yaml` file indicates whether a bootloader will be included in the volume and, if so, the type of bootloader it will be.

You can build classical images with ubuntu-image using the following command:

```bash
ubuntu-image classic image_definition.yaml
```

> For more information on how to build a gadget snap, visit [Building a gadget snap](https://ubuntu.com/core/docs/gadget-building).

Reviewer: Considering we have a section on building classical images, it seems strange that snap-based images don't have their own section.

How about having just a generic section called "Building images", which would only include very basic information and then links to instructions on how to do the actual building?

## Build and test ubuntu-image

Visit [contributing to ubuntu-image](https://github.com/canonical/ubuntu-image/blob/main/CONTRIBUTING.md) for instructions on how to set up, build, and test ubuntu-image in development mode.

## License

> The ubuntu-image project is licensed under [GNU General Public License v3.0](https://github.com/canonical/ubuntu-image/blob/main/LICENSE). So, you are free to use, modify, and redistribute it without further restrictions.

Reviewer: The second sentence is a bit superfluous (and potentially misleading).

```suggestion
The ubuntu-image project is licensed under [GNU General Public License v3.0](https://github.com/canonical/ubuntu-image/blob/main/LICENSE).
```

> ## Contribute to ubuntu-image

Reviewer: ```suggestion
## Contributing to ubuntu-image
```

> To contribute to the ubuntu-image project, visit [contributing to ubuntu-image](https://github.com/canonical/ubuntu-image/blob/main/CONTRIBUTING.md).

Reviewer: ```suggestion
To learn how to contribute to the ubuntu-image project, visit [Contributing to ubuntu-image](https://github.com/canonical/ubuntu-image/blob/main/CONTRIBUTING.md).
```

# ubuntu-image

[![ubuntu-image](https://snapcraft.io/ubuntu-image/badge.svg)](https://snapcraft.io/ubuntu-image)
![Build](https://github.com/canonical/ubuntu-image/actions/workflows/build-and-test.yml/badge.svg)
[![codecov](https://codecov.io/gh/canonical/ubuntu-image/branch/main/graph/badge.svg?token=F9jE9HKo1a)](https://codecov.io/gh/canonical/ubuntu-image)
[![Go Report Card](https://goreportcard.com/badge/github.com/canonical/ubuntu-image)](https://goreportcard.com/report/github.com/canonical/ubuntu-image)

ubuntu-image is a tool used for generating bootable images. You can use it to build Ubuntu images such as:

- Snap-based Ubuntu Core images from model assertions

- Classical preinstalled Ubuntu images using image definitions

The future versions of this tool will be more generalized, allowing users to build a wider range of Ubuntu images, including all future releases.

## Getting started

### Requirements

> To get started, you will need Ubuntu 18.04 (Bionic Beaver) as a minimum platform requirement. However, it is recommended to use Ubuntu 20.04 (Focal Fossa) or newer. All necessary third-party packages are available in the Ubuntu archive.

Reviewer: Let's make it as concise as possible:
```suggestion
* Ubuntu 18.04 (Bionic Beaver) or newer (recommended: Ubuntu 20.04 (Focal Fossa) or newer)
* Ability to install snaps ([SnapStore: ubuntu-image](https://snapcraft.io/ubuntu-image))
```

### Quickstart

See [Build your first Ubuntu Core image](https://ubuntu.com/core/docs/build-an-image) for instructions on how to build an Ubuntu core image on a **Raspberry Pi**.

### Building images

ubuntu-image offers two basic sub-commands for building snap-based and classical images.

> To build a snap-based image with ubuntu-image, you need a [model assertion](https://ubuntu.com/core/docs/reference/assertions/model?_ga=2.79789975.141992633.1711553306-121934327.1710189436). A model assertion is a YAML file that describes a particular combination of core, kernel, and gadget snaps, along with other declarations, signed with a digital signature asserting its authenticity. The `ubuntu-image` command only requires the path to this model assertion to build snap-based images.

Reviewer: ```suggestion
To build a snap-based image with ubuntu-image, you need a [model assertion](https://ubuntu.com/core/docs/reference/assertions/model). A model assertion is a YAML file that describes a particular combination of core, kernel, and gadget snaps, along with other declarations, signed with a digital signature asserting its authenticity. The `ubuntu-image` command only requires the path to this model assertion to build snap-based images.
```

The `gadget.yaml` file contains important information about the volumes that will be created. You can think of these volumes as disk images. The file also specifies the structures that will be included within each volume. Structures define the layout of the volume, including partitions, Master Boot Records, and any other relevant content. Also, the `gadget.yaml` file indicates whether a bootloader will be included in the volume and, if so, the type of bootloader it will be.

To build classical images with ubuntu-image using the following command:

```
ubuntu-image classic image_definition.yaml
```

For more information on how to build a gadget snap, see [Building a gadget snap](https://ubuntu.com/core/docs/gadget-building).

## Building and testing ubuntu-image

> See [contributing to ubuntu-image](https://github.com/canonical/ubuntu-image/blob/main/CONTRIBUTING.md) for instructions on how to set up, build, and test ubuntu-image in development mode.

Reviewer: ```suggestion
See [Contributing to ubuntu-image](https://github.com/canonical/ubuntu-image/blob/main/CONTRIBUTING.md) for instructions on how to set up, build, and test ubuntu-image in development mode.
```

# ubuntu-image

[![ubuntu-image](https://snapcraft.io/ubuntu-image/badge.svg)](https://snapcraft.io/ubuntu-image)
![Build](https://github.com/canonical/ubuntu-image/actions/workflows/build-and-test.yml/badge.svg)
[![codecov](https://codecov.io/gh/canonical/ubuntu-image/branch/main/graph/badge.svg?token=F9jE9HKo1a)](https://codecov.io/gh/canonical/ubuntu-image)
[![Go Report Card](https://goreportcard.com/badge/github.com/canonical/ubuntu-image)](https://goreportcard.com/report/github.com/canonical/ubuntu-image)

ubuntu-image is a tool used for generating bootable images. You can use it to build Ubuntu images such as:

- Snap-based Ubuntu Core images from model assertions

- Classical preinstalled Ubuntu images using image definitions

> The future versions of this tool will be more generalized, allowing users to build a wider range of Ubuntu images, including all future releases.

Reviewer 2: "all ubuntu images" meant "all 'kind' of ubuntu images, so ISO/installer too", not "all future releases". Does it make sense?

## Getting started

### Requirements

> * Ubuntu 18.04 (Bionic Beaver) or newer (recommended: Ubuntu 20.04 (Focal Fossa) or newer)

Reviewer 2: ```suggestion
* Ubuntu 18.04 (Bionic Beaver) or newer (recommended: Ubuntu 22.04 (Jammy Jellyfish) or newer)
```

I guess we should recommend the latest LTS.

This is weird to have "or newer" twice, no?

* Ability to install snaps ([SnapStore: ubuntu-image](https://snapcraft.io/ubuntu-image))

### Quickstart

See [Build your first Ubuntu Core image](https://ubuntu.com/core/docs/build-an-image) for instructions on how to build an Ubuntu core image on a **Raspberry Pi**.

> ### Building images

Reviewer 2: Maybe use two sub-sections for the two modes of operation? Because in the final markdown render, it looks a bit hard to distinguish the two blocks of instructions.

ubuntu-image offers two basic sub-commands for building snap-based and classical images.

To build a snap-based image with ubuntu-image, you need a [model assertion](https://ubuntu.com/core/docs/reference/assertions/model). A model assertion is a YAML file that describes a particular combination of core, kernel, and gadget snaps, along with other declarations, signed with a digital signature asserting its authenticity. The `ubuntu-image` command only requires the path to this model assertion to build snap-based images.

To build snap-based images with `ubuntu-image`, use the following command:

```
ubuntu-image snap model.assertion
```

See [Build your first Ubuntu Core image](https://ubuntu.com/core/docs/build-an-image) for more information on building snap-based images using ubuntu-image. To build an image with custom snaps, see [Build an image with custom snaps](https://ubuntu.com/core/docs/custom-images).

> Classical images are built from a local [gadget tree](https://github.com/snapcore/pc-gadget) path. The [gadget tree](https://github.com/snapcore/pc-gadget) is a primed [gadget snap](https://snapcraft.io/docs/the-gadget-snap) containing a `gadget.yaml` file in the meta directory and all the necessary bootloader gadget bits built. You can prepare a [gadget tree](https://github.com/snapcore/pc-gadget) by fetching a specially customised [gadget snap](https://snapcraft.io/docs/the-gadget-snap) source and running `snapcraft prime` on it. The resulting tree will be stored in the prime/ directory.

Reviewer 2: I know it comes from [this doc](https://canonical-subiquity.readthedocs-hosted.com/en/latest/reference/ubuntu-image.html) but (you cannot know that @activus-d so no worries!) I really find this confusing to focus a whole paragraph here on something that we do not really want our users to do because building the gadget tree is now done by ubuntu-image if we specify a path to a gadget.yaml file in the image definition YAML. The *most important* input for a classic build is the image definition YAML, currently best documented [here](https://github.com/canonical/ubuntu-image/blob/main/internal/imagedefinition/README.rst) (but we would like that to move). So I think we should focus on that and make it easy to find this doc to easily get started.

The `gadget.yaml` file contains important information about the volumes that will be created. You can think of these volumes as disk images. The file also specifies the structures that will be included within each volume. Structures define the layout of the volume, including partitions, Master Boot Records, and any other relevant content. Also, the `gadget.yaml` file indicates whether a bootloader will be included in the volume and, if so, the type of bootloader it will be.

To build classical images with ubuntu-image, use the following command:

```
ubuntu-image classic image_definition.yaml
```

> For more information on how to build a gadget snap, see [Building a gadget snap](https://ubuntu.com/core/docs/gadget-building).

Reviewer 2: A bit weird to get a link to that but nothing about the mandatory image definition file, which is even more important because many users will reuse existing gadgets but will write original image definition.

## Building and testing ubuntu-image

See [Contributing to ubuntu-image](https://github.com/canonical/ubuntu-image/blob/main/CONTRIBUTING.md) for instructions on how to set up, build, and test ubuntu-image in development mode.

## License

The ubuntu-image project is licensed under [GNU General Public License v3.0](https://github.com/canonical/ubuntu-image/blob/main/LICENSE).

## Contributing to ubuntu-image

> To learn how to contribute to the ubuntu-image project, see [Contributing to ubuntu-image](https://github.com/canonical/ubuntu-image/blob/main/CONTRIBUTING.md).

Reviewer 2: Any particular reason to use full URLs instead of relative paths?

Author: @upils no particular reason that I can think of. I can use relative paths if that won't adversely affect the links on [pkg.go.dev/github.com/canonical/ubuntu-image](https://pkg.go.dev/github.com/canonical/ubuntu-image).

Reviewer: I expected relative links wouldn't work on pkg.go.dev.

The `gadget.yaml` file contains important information about the volumes that will be created. You can think of these volumes as disk images. The file also specifies the structures that will be included within each volume. Structures define the layout of the volume, including partitions, Master Boot Records, and any other relevant content. Also, the `gadget.yaml` file indicates whether a bootloader will be included in the volume and, if so, the type of bootloader it will be.

> To build classical images with ubuntu-image use the following command:

Reviewer: ```suggestion
To build classical images with ubuntu-image, use the following command:
```
